core-rules/src/main/java/org/hisp/dhis/android/rules/RuleEngine.java

```diff
@@ -1,112 +1,125 @@
 package org.hisp.dhis.android.rules;
 
-import org.hisp.dhis.android.rules.models.Event;
-import org.hisp.dhis.android.rules.models.ProgramRule;
-import org.hisp.dhis.android.rules.models.RuleVariable;
+import org.hisp.dhis.android.rules.models.RuleEffect;
+import org.hisp.dhis.android.rules.models.RuleEnrollment;
+import org.hisp.dhis.android.rules.models.RuleEvent;
 
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
+import java.util.Locale;
 import java.util.concurrent.Callable;
 
 import javax.annotation.Nonnull;
-
-import static java.util.Collections.unmodifiableList;
+import javax.annotation.Nullable;
 
 
 public final class RuleEngine {
-    private final ExpressionEvaluator evaluator;
-    private final List<ProgramRule> programRules;
-    private final List<RuleVariable> ruleVariables;
-    private final List<Event> events;
-
-    RuleEngine(ExpressionEvaluator evaluator,
-            List<ProgramRule> programRules,
-            List<RuleVariable> ruleVariables,
-            List<Event> events) {
-        this.evaluator = evaluator;
-        this.programRules = programRules;
-        this.ruleVariables = ruleVariables;
-        this.events = events;
-    }
 
-    public static Builder builder(@Nonnull ExpressionEvaluator evaluator) {
-        if (evaluator == null) {
-            throw new IllegalArgumentException("evaluator == null");
-        }
+    @Nonnull
+    private final RuleEngineContext ruleEngineContext;
+
+    @Nonnull
+    private final List<RuleEvent> ruleEvents;
 
-        return new Builder(evaluator);
+    @Nullable
+    private final RuleEnrollment ruleEnrollment;
+
+    RuleEngine(@Nonnull RuleEngineContext ruleEngineContext,
+            @Nonnull List<RuleEvent> ruleEvents,
+            @Nullable RuleEnrollment ruleEnrollment) {
+        this.ruleEngineContext = ruleEngineContext;
+        this.ruleEvents = ruleEvents;
+        this.ruleEnrollment = ruleEnrollment;
     }
 
     @Nonnull
-    public List<ProgramRule> programRules() {
-        return programRules;
+    public List<RuleEvent> events() {
+        return ruleEvents;
+    }
+
+    @Nullable
+    public RuleEnrollment enrollment() {
+        return ruleEnrollment;
     }
 
     @Nonnull
-    public List<RuleVariable> programRuleVariables() {
-        return ruleVariables;
+    public RuleEngineContext executionContext() {
+        return ruleEngineContext;
     }
 
     @Nonnull
-    public List<Event> events() {
-        return events;
+    public Callable<List<RuleEffect>> evaluate(@Nonnull RuleEvent ruleEvent) {
+        if (ruleEvent == null) {
+            throw new IllegalArgumentException("ruleEvent == null");
+        }
+
+        for (RuleEvent contextualEvent : ruleEvents) {
```

Comment:
There is a possibility that client will send ruleEvent with different fields to evaluate(), so we can't rely on .equals() or .hashcode() in this case.